sqldb/invoices.go

```diff
@@ -0,0 +1,1651 @@
+			}
+		}
+
+		invoice.AMPState = ampState
+		invoice.Htlcs = ampHtlcs
+
+		return hash, invoice, nil
+	}
+
+	// Otherwise simply fetch the invoice HTLCs.
+	htlcs, err := getInvoiceHtlcs(ctx, db, row.ID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(htlcs) > 0 {
+		invoice.Htlcs = htlcs
+		var amountPaid lnwire.MilliSatoshi
+		for _, htlc := range htlcs {
+			if htlc.State == invpkg.HtlcStateSettled {
+				amountPaid += htlc.Amt
+			}
+		}
+		invoice.AmtPaid = amountPaid
+	}
+
+	return hash, invoice, nil
+}
+
+// getInvoiceFeatures fetches the invoice features for the given invoice id.
+func getInvoiceFeatures(ctx context.Context, db InvoiceQueries,
+	invoiceID int64) (*lnwire.FeatureVector, error) {
+
+	rows, err := db.GetInvoiceFeatures(ctx, invoiceID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get invoice features: %w",
+			err)
+	}
+
+	features := lnwire.EmptyFeatureVector()
+	for _, feature := range rows {
+		features.Set(lnwire.FeatureBit(feature.Feature))
+	}
+
+	return features, nil
+}
+
+// getInvoiceHtlcs fetches the invoice htlcs for the given invoice id.
+func getInvoiceHtlcs(ctx context.Context, db InvoiceQueries,
+	invoiceID int64) (map[invpkg.CircuitKey]*invpkg.InvoiceHTLC, error) {
+
+	htlcRows, err := db.GetInvoiceHTLCs(ctx, invoiceID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get invoice htlcs: %w", err)
+	}
+
+	// We have no htlcs to unmarshal.
+	if len(htlcRows) == 0 {
+		return nil, nil
+	}
+
+	crRows, err := db.GetInvoiceHTLCCustomRecords(ctx, invoiceID)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get custom records for "+
+			"invoice htlcs: %w", err)
+	}
+
+	cr := make(map[int64]record.CustomSet, len(crRows))
+	for _, row := range crRows {
+		if _, ok := cr[row.HtlcID]; !ok {
+			cr[row.HtlcID] = make(record.CustomSet)
+		}
+
+		value := row.Value
+		if value == nil {
+			value = []byte{}
+		}
+		cr[row.HtlcID][uint64(row.Key)] = value
+	}
+
+	htlcs := make(map[invpkg.CircuitKey]*invpkg.InvoiceHTLC, len(htlcRows))
+
+	for _, row := range htlcRows {
+		circuiteKey, htlc, err := unmarshalInvoiceHTLC(row)
+		if err != nil {
+			return nil, fmt.Errorf("unable to unmarshal "+
+				"htlc(%d): %w", row.ID, err)
+		}
+
+		if customRecords, ok := cr[row.ID]; ok {
+			htlc.CustomRecords = customRecords
+		} else {
+			htlc.CustomRecords = make(record.CustomSet)
+		}
+
+		htlcs[circuiteKey] = htlc
+	}
+
+	return htlcs, nil
+}
+
+// unmarshalInvoice converts an InvoiceRow to an Invoice.
+func unmarshalInvoice(row sqlc.Invoice) (*lntypes.Hash, *invpkg.Invoice,
+	error) {
+
+	var (
+		settledIndex   int64
+		settledAt      time.Time
+		memo           []byte
+		paymentRequest []byte
+		preimage       *lntypes.Preimage
+		paymentAddr    [32]byte
+	)
+
+	hash, err := lntypes.MakeHash(row.Hash)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if row.SettledIndex.Valid {
+		settledIndex = row.SettledIndex.Int64
+	}
+
+	if row.SettledAt.Valid {
+		settledAt = row.SettledAt.Time.Local()
```

Comment:
Nit: UTC ?

Reply:
We need to use local time to be compatible with channeldb, see my other comment.